src/Storages/MergeTree/MergeTreeReadPoolParallelReplicasInOrder.cpp

```diff
@@ -65,21 +71,77 @@ MergeTreeReadTaskPtr MergeTreeReadPoolParallelReplicasInOrder::getTask(size_t ta
 
     if (task_idx >= per_part_infos.size())
-        throw Exception(ErrorCodes::LOGICAL_ERROR,
-            "Requested task with idx {}, but there are only {} parts",
-            task_idx, per_part_infos.size());
+        throw Exception(
+            ErrorCodes::LOGICAL_ERROR, "Requested task with idx {}, but there are only {} parts", task_idx, per_part_infos.size());
 
     const auto & part_info = per_part_infos[task_idx]->data_part->info;
-    auto get_from_buffer = [&]() -> std::optional<MarkRanges>
+    const auto & data_settings = per_part_infos[task_idx]->data_part->storage.getSettings();
+    auto get_from_buffer = [&,
+                            rows_granularity = (*data_settings)[MergeTreeSetting::index_granularity],
+                            my_max_block_size = this->block_size_params.max_block_size_rows]() -> std::optional<MarkRanges>
     {
+        const size_t max_marks_in_range = (my_max_block_size + rows_granularity - 1) / rows_granularity;
         for (auto & desc : buffered_tasks)
         {
             if (desc.info == part_info && !desc.ranges.empty())
             {
-                auto result = std::move(desc.ranges);
-                desc.ranges = MarkRanges{};
-                ProfileEvents::increment(ProfileEvents::ParallelReplicasReadMarks, desc.ranges.getNumberOfMarks());
-                return result;
+                if (mode == CoordinationMode::WithOrder)
+                {
+                    /// if already splited, just return desc.ranges
+                    if (split)
+                    {
+                        auto result = std::move(desc.ranges);
+                        desc.ranges = MarkRanges{};
+                        ProfileEvents::increment(ProfileEvents::ParallelReplicasReadMarks, result.getNumberOfMarks());
+                        return result;
+                    }
+
+                    /// else split range but still return all MarkRanges
+                    MarkRanges result;
+                    for (auto range : desc.ranges)
+                    {
+                        while (!split && range.begin + marks_in_range < range.end)
+                        {
+                            result.emplace_back(range.begin, range.begin + marks_in_range);
+                            range.begin += marks_in_range;
+                            marks_in_range *= 2;
+                            if (marks_in_range > max_marks_in_range)
+                                split = true;
+                        }
+                        result.emplace_back(range.begin, range.end);
+                    }
+                    chassert(!result.empty());
+                    desc.ranges = MarkRanges{};
+                    ProfileEvents::increment(ProfileEvents::ParallelReplicasReadMarks, result.getNumberOfMarks());
+                    return result;
+                }
+                else
+                {
+                    /// for reverse order just return one range
```

Comment:
i don't understand why

Reply:
In `MergeTreeInReverseOrderSelectAlgorithm::readFromTask`, clickhouse will read all data in `MergeTreeReadTask`. 

https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreeSelectAlgorithms.cpp#L45

So I think if we still return all MarkRanges to create a task, `MergeTreeInReverseOrderSelectAlgorithm` will still store the whole ranges in memory. So i just return one range to create a task

Reply:
maybe it is ok if we do it once for each part (task with 1 mark, then with 2, then with 4 until we reach `min_marks_per_task`), but then we shouldn't reset `marks_in_range` back to 1